### ISSUE TITLE
Feature Request: Activation offloading with async prefetch in FSDP

### ISSUE THREAD

**Issue by niyunsheng** [pytorch/pytorch#174960]
### Motivation

FSDP already manages **parameter CPU offloading** via `CPUOffloadPolicy`: parameters are offloaded to CPU when not needed, and loaded back to GPU before compute. The prefetch scheduling is handled automatically by FSDP's runtime — in backward, it calls `unshard()` (which includes H2D + all-gather) one layer ahead by default, and users can further customize via `set_modules_to_backward_prefetch()`.

Activation tensors have an identical lifecycle: produced during forward, not needed until backward, and can be offloaded to CPU in the meantime. Moreover, activation offloading is simpler than parameter offloading — it only requires H2D/D2H copies (local PCIe), no all-gather (cross-rank communication). I propose adding activation offloading as a first-class feature within `CPUOffloadPolicy`, reusing FSDP's existing prefetch infrastructure.

### Why `CPUOffloadPolicy` is the right place

Parameter CPU offload and activation offload follow the same pattern:

| | Parameter CPU offload (today) | Activation CPU offload (proposed) |
| --- | --- | --- |
| **Offload** | After forward, D2H via PCIe | After a layer's forward, D2H via PCIe |
| **Load** | Before compute, H2D via PCIe | Before backward, H2D via PCIe |
| **Prefetch** | `_backward_prefetch`: unshard next layer ahead | Same: load next layer's activations ahead |
| **Communication** | H2D + all-gather | H2D only (simpler) |

Placing both in `CPUOffloadPolicy` enables:

1. **Unified PCIe scheduling**: Parameter H2D and activation H2D share PCIe bandwidth. A single framework can coordinate both to avoid contention.
2. **Reuse prefetch infrastructure**: FSDP already tracks module execution order and implements `_backward_prefetch` (default 1 layer ahead). Activation prefetch can plug into the same mechanism — trigger activation H2D in `pre_backward`, just as parameter `unshard()` is triggered there today.
3. **Consistent API**: No new prefetch parameters needed. Activation offload follows the same defaults and manual controls as parameter offload.

### Proposed API

```python
fully_shard(module, offload_policy=CPUOffloadPolicy(
    offload_activations=True,
))
```

No separate prefetch parameter — activation prefetch reuses FSDP's existing scheduling:
* Default: prefetch 1 layer ahead in backward (same as parameter prefetch)
* Manual control: via set_modules_to_backward_prefetch() or a similar API if independent tuning is needed

### Implementation sketch

Activation offload integrates into FSDP's existing hook system:
* `post_forward`: After a layer's forward, async D2H copy activations to CPU pinned memory (via a separate CUDA stream)
* pre_backward: Like `_backward_prefetch` calls `unshard()` for the next layer's parameters, also trigger H2D for the next layer's activations
* Ping-pong GPU buffers: Two alternating buffers — one serves backward compute while the other receives the next prefetch

### Key considerations

* **Pre-allocated pinned memory pool**: Dynamic pinned memory allocation triggers CUDA implicit synchronization, preventing actual stream overlap (see https://github.com/meta-pytorch/torchtune/issues/2076). Pre-allocated buffers are needed.
* **Orthogonal to checkpointing**: Activation offloading preserves activations (on CPU) while checkpointing discards them (and recomputes). They are independent features that can be composed: offload alone (no recompute), checkpoint alone, or both together.

### Prior art

* [torchtune `OffloadActivations`](https://github.com/meta-pytorch/torchtune/blob/main/torchtune/training/_activation_offloading.py): Standalone offloading via `saved_tensors_hooks`. Has known stream sync issues from dynamic pinned memory allocation 1.
* [Megatron-Kwai `OffloadManager`](https://github.com/kwai/Megatron-Kwai/blob/sc25slimpipe/megatron/core/pipeline_parallel/offload.py): Batch-level offload/onload with ping-pong GPU buffers, pre-allocated pinned memory, and separate D2H/H2D streams.

### Related

* #158657 — proposes extending `torch.utils.checkpoint` with offloading (different design direction; couples offloading with recomputation)

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @janeyx99

**Comment by weifengpy** [comment]
@niyunsheng is activation offloading always happening at fully_shard moduel boundary? What about interemediate activations/tensors inside the layer? in the following example, fully_shard can only offload output_tensor but not intermediate_tensor. Is this a valid question?

```
fully_shard(model)

def forward(self, input):
    intermediate_tensor = torch.mm(input, input)
    ...
    return output_tensor
```

**Comment by niyunsheng** [comment]
@weifengpy Thanks for the positive feedback! Addressing your questions:

> what's your opinion on this two approach: 1) extending torch.utils.checkpoint with offloading, 2) extending fully_shard for offloading (this proposal). Pros and cons?

I believe they serve **different scenarios** and are complementary:

- **AC + offload (extending `checkpoint`, #158657)**: For shorter sequences where PCIe bandwidth may not fully hide all transfers. In this case, using AC to recompute cheap ops and offloading only the expensive ones (GEMM/FA outputs) is reasonable. @soulitzer's proposed `CheckpointPolicy.OFFLOAD` in SAC handles this well.

- **Offload-only (extending `fully_shard`, this proposal)**: For long-sequence training where offloading everything without any checkpoint is optimal. Here's the analysis: for a GEMM `[S,H]×[H,H]`, the activation saved for backward is the input `[S,H]`. On H200 (989 TFLOPS bf16, 64 GB/s PCIe Gen5), the ratio of offload time to compute time for a single GEMM is `989000/(64*H)` ≈ 3.8x for H=4096 — one GEMM alone can't hide its activation offload. But flash attention's FLOPs are **2S/H times** a single GEMM. At S=128K, H=4096, FA is **64x** the compute of one GEMM. A transformer layer has 6-8 GEMMs plus FA, and the offload of layer `L` overlaps with layer `L+1`'s forward. In the long-sequence regime, FA dominates per-layer compute, providing ample time to hide all PCIe transfers — eliminating **all** redundant recomputation including memory-bound ops (layernorm, activations, dropout) that selective AC still re-executes inside checkpointed blocks.

Both approaches require AC and offload to be **independently composable**, which supports
designing them as orthogonal features.

> are you working with some of us direclty already? is this confirming before proceeding or want us to implement this?

I'm not working with anyone on the PyTorch team directly on this. I have experience implementing
async activation offloading based on the [Megatron-Kwai approach](https://github.com/kwai/Megatron-Kwai/blob/sc25slimpipe/megatron/core/pipeline_parallel/offload.py) and am familiar with the key engineering challenges (stream synchronization, pinned memory management, ping-pong buffers). 

I'm happy to contribute a torchtitan-level implementation first as an interim solution (see [torchtitan#2379](https://github.com/pytorch/torchtitan/issues/2379), which has been marked high priority and assigned to @soulitzer). 

But ideally, I think this should be natively supported in FSDP's `CPUOffloadPolicy` so that all downstream frameworks can benefit without each reimplementing their own offloading logic.

> do you want to use this feature through torchtitan or directly pytorch fully_shard with your own training loop?

I primarily use torchtitan. For the short term, a torchtitan-level implementation using `saved_tensors_hooks` is practical without PyTorch core changes. 

For the long term, I believe this should live in `fully_shard` / `CPUOffloadPolicy` for the reasons in this proposal (unified PCIe scheduling, reusing prefetch infrastructure).

> is activation offloading always happening at `fully_shard` module boundary? What about intermediate activations/tensors inside the layer?

Good question. If implemented via `saved_tensors_hooks`, the hooks intercept **all tensors saved by autograd for backward** — not just the module-level output. This includes intermediate tensors inside the layer (e.g. MLP intermediate results). 
So offloading naturally covers all activations, not limited to module boundaries.

This also illustrates why offload is orthogonal to checkpoint: when checkpoint is enabled, it discards most saved tensors (and recomputes them); when offload is enabled, it moves all saved tensors to CPU (no recompute). When both are enabled, offload handles whatever tensors checkpoint decides to keep.

**Comment by ngimel** [comment]
>providing ample time to hide all PCIe transfers — eliminating all redundant recomputation including memory-bound ops (layernorm, activations, dropout) that selective AC still re-executes inside checkpointed blocks.

CPU memory and PCIe bandwidth is not an infinite resource, and it never makes sense to offload (at 65 GB/s) things that can be recomputed at 3 TB/s, regardless of how much compute you have to cover.

**Comment by niyunsheng** [comment]
> > providing ample time to hide all PCIe transfers — eliminating all redundant recomputation including memory-bound ops (layernorm, activations, dropout) that selective AC still re-executes inside checkpointed blocks.
> 
> CPU memory and PCIe bandwidth is not an infinite resource, and it never makes sense to offload (at 65 GB/s) things that can be recomputed at 3 TB/s, regardless of how much compute you have to cover.

@ngimel I partially agree with the bandwidth constraint analysis. Offloading all activations consumes PCIe bandwidth and may cause contention with other asynchronous operations, which can lead to potential slowdowns. For memory-bound operators (e.g., norm, RoPE), recomputation utilizing high HBM bandwidth is generally more efficient.

However, empirical profiling indicates that pure activation offloading still yields a marginal performance improvement (~0.5%) in specific long-sequence scenarios. Detailed model configurations and profiling metrics will be provided subsequently.

The primary architectural consideration is that activation offloading and activation checkpointing (AC) are fundamentally orthogonal primitives. They should not be coupled within the framework implementation.
1. System Flexibility: Users must be permitted to configure pure activation offloading without being forced to enable AC or selective recomputation.
2. Architectural Variance: Non-standard or future model architectures may possess different ratios of memory-bound to compute-bound operators, shifting the offload vs. recompute break-even point.
3. Composability: Decoupling allows developers to independently tune combinations of offloading and recomputation based on their specific hardware topologies and PCIe-to-HBM bandwidth ratios.

Providing offloading as an independent capability ensures the necessary infrastructural flexibility.

**Comment by rakkit** [comment]
@niyunsheng 
> should natively saturate this bandwidth without additional abstractions

i hope so but i assume we still need scheduler to make sure pre-fetch works properly.

and i found the test i did above is really times ago in junly 2024, so maybe need rerun some test on latest version

**Comment by MatthiasKohl** [comment]
@ngimel depending on the system, there are valid use-cases where recomputation is slower than offloading, see https://www.nvidia.com/en-us/on-demand/session/gtc24-s61159/

Even with PCIe, if computation is heavily compute-bound, the required bandwidth might be below PCIe bandwidth, thus offloading can be worth it. With the numbers you are assuming (65 GB/s PCIe BW, 3 TB/s HBM BW), no common/popular model will be in this regime, so you are correct in saying that it rarely makes sense to offload, but there can still be use-cases which fall into this category.

Plus, it also depends on the inter-connect between devices: if it is comparable in BW to PCIe BW, then offloading might enable running a larger model on the same device, which is typically faster than distributing the larger model to multiple devices due to additional overheads for this (and no gain in BW as we assume BW between devices is in the same ballpark as PCIe). If offloading is worth it even if we are bound by PCIe BW, asynchronous offloading is generally still faster / provides a better trade-off between speed and memory usage.

**Comment by ngimel** [comment]
@MatthiasKohl I was responding to a comment about offloading  "memory-bound ops (layernorm, activations, dropout)" - they are not heavily compute bound. I agree that offloading expensive computation is useful.
